crates/autopilot/src/run_loop.rs

```diff
@@ -389,6 +400,89 @@ impl RunLoop {
         .collect()
     }
 
+    /// Returns true if winning solution is fair or winner is None
+    fn is_solution_fair(
+        winner: Option<&Participant>,
+        remaining: &Vec<Participant>,
+        auction: &domain::Auction,
+    ) -> bool {
+        let Some(winner) = winner else { return true };
+        let Some(fairness_threshold) = winner.driver.fairness_threshold else {
+            return true;
+        };
+
+        // Returns the surplus difference in the buy token if `left`
+        // is better for the trader than `right`, or 0 otherwise.
+        // This takes differently partial fills into account.
+        let improvement_in_buy = |left: &TradedAmounts, right: &TradedAmounts| {
+            // If `left.sell / left.buy < right.sell / right.buy`, left is "better" as the
+            // trader either sells less or gets more. This can be reformulated as
+            // `right.sell * left.buy > left.sell * right.buy`.
+            let right_sell_left_buy = right.sell.0.full_mul(left.buy.0);
+            let left_sell_right_buy = left.sell.0.full_mul(right.buy.0);
+            let improvement = right_sell_left_buy
+                .checked_sub(left_sell_right_buy)
+                .unwrap_or_default();
+
+            // The difference divided by the original sell amount is the improvement in buy
+            // token. Casting to U256 is safe because the difference is smaller than the
+            // original product, which if re-divided by right.sell must fit in U256.
+            improvement
+                .checked_div(right.sell.0.into())
+                .map(|v| U256::try_from(v).expect("improvement in buy fits in U256"))
+                .unwrap_or_default()
+        };
+
+        // Find best execution per order
+        let mut best_executions = HashMap::new();
+        for other in remaining {
+            for (uid, execution) in other.solution.orders() {
+                let best_execution = best_executions.entry(uid).or_insert(execution);
+                if !improvement_in_buy(execution, best_execution).is_zero() {
+                    *best_execution = execution;
+                }
```

Comment:
nit: avoid searching for improvement with self.
```suggestion
                best_executions
                    .entry(uid)
                    .and_modify(|best_execution| {
                        if !improvement_in_buy(execution, best_execution).is_zero() {
                            *best_execution = execution;
                        }
                    })
                    .or_insert(execution);
```